src/CodeFormatter.cs

```diff
@@ -293,99 +289,105 @@ private static void LogWorkspaceDiagnostics(ILogger logger, bool logWorkspaceWar
-                getDocumentsAndOptions.AddRange(getProjectDocuments);
-            }
+                foreach (var document in project.Documents)
+                {
+                    // If we've already added this document, either via a link or multi-targeted framework, then ignore.
+                    if (document?.FilePath is null ||
+                        addedFilePaths.Contains(document.FilePath))
+                    {
+                        continue;
+                    }
 
-            var documentsAndOptions = await Task.WhenAll(getDocumentsAndOptions).ConfigureAwait(false);
-            var foundEditorConfig = documentsAndOptions.Any(documentAndOptions => documentAndOptions.Item4);
+                    addedFilePaths.Add(document.FilePath);
 
-            var addedFilePaths = new HashSet<string>(documentsAndOptions.Length);
-            var formattableFiles = ImmutableArray.CreateBuilder<(DocumentId, OptionSet, ICodingConventionsSnapshot)>(documentsAndOptions.Length);
-            foreach (var (document, options, codingConventions, hasEditorConfig) in documentsAndOptions)
-            {
-                if (document?.FilePath is null)
-                {
-                    continue;
-                }
+                    if (!fileMatcher.Match(document.FilePath).HasMatches ||
+                        !document.SupportsSyntaxTree)
+                    {
+                        continue;
+                    }
 
-                // If any code file has an .editorconfig, then we should ignore files without an .editorconfig entry.
-                if (foundEditorConfig && !hasEditorConfig)
-                {
-                    continue;
-                }
+                    var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                    if (syntaxTree is null)
+                    {
+                        throw new Exception($"Unable to get a syntax tree for '{document.Name}'");
+                    }
 
-                // If we've already added this document, either via a link or multi-targeted framework, then ignore.
-                if (addedFilePaths.Contains(document.FilePath))
-                {
-                    continue;
+                    if (!includeGeneratedFiles &&
+                        await GeneratedCodeUtilities.IsGeneratedCodeAsync(syntaxTree, cancellationToken).ConfigureAwait(false))
+                    {
+                        continue;
+                    }
+
+                    var analyzerConfigOptions = await GetAnalyzerConfigOptionsAsync(document, cancellationToken).ConfigureAwait(false);
+                    var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+                    var formattableDocument = new DocumentWithOptions(document, optionSet, analyzerConfigOptions);
```

Comment:
Why is DocumentWithOptions needed if we can fetch the OptionSet when you need it later or potentially the AnalyzerConfigOptions too? Is this just a perf optimization or needed for something else?

Reply:
I thought I was saving work, but you are right. Since nothing in the solution is changing and these are cached within the DocumentState and AnalyzerConfigSet respectively. This could just return document ids and we could fetch everything again later.

Reply:
I'll take this suggestion as a follow up refactoring. https://github.com/dotnet/format/issues/691